FE: Only allow enabled factors to be set up in API

AnthonyKot opened the pull request:
# Which Problems Are Solved

- Hides for users MFA options are not allowed by org policy.
- Fix for "ng test" across "console"

# How the Problems Are Solved

- Before displaying MFA options we call "listMyMultiFactors" from parent component to filter MFA allowed by org

# Additional Changes

- Dependency Injection was fixed around ng unit tests

# Additional Context

admin view
<img width="698" alt="Screenshot 2025-02-06 at 00 26 50" src="https://github.com/user-attachments/assets/1b642c8a-a640-4bdd-a1ca-bde70c263567" />
user view
<img width="751" alt="Screenshot 2025-02-06 at 00 27 16" src="https://github.com/user-attachments/assets/e1c99907-3226-46ce-b8bc-e993af4b4cae" />
test
<img width="1500" alt="Screenshot 2025-02-06 at 00 01 36" src="https://github.com/user-attachments/assets/3b742655-6cbb-4bc9-b09b-b859393bfbe4" />

The issue: https://github.com/zitadel/zitadel/issues/9176
The bug report: https://discord.com/channels/927474939156643850/1307006457815896094

muhlemmer commented:
Hi thanks for the PR. We are currently reviewing a [PR](https://github.com/zitadel/zitadel/pull/9312) which refactors some console code. This **might** conflict with some of the work in this PR. We will defer the review until the other is merged, as it has priority for our road map.

We will get back as soon as that's done.

AnthonyKot commented:
Hey @muhlemmer, rebased on master after the PR was merged

muhlemmer commented:
CC @conblem

conblem commented:
Thanks @AnthonyKot for your patience, im taking a look at your PR today

conblem commented on console/src/app/pages/users/user-detail/auth-user-detail/auth-user-mfa/auth-user-mfa.component.ts:
```diff
@@ -59,6 +65,10 @@ export class AuthUserMfaComponent implements OnInit, OnDestroy {
   public addAuthFactor(): void {
     const dialogRef = this.dialog.open(AuthFactorDialogComponent, {
       data: {
```
It would be nice if you could add typing to the dialog input. This involves creating a Data and Result type on the dialog like here: https://github.com/zitadel/zitadel/blob/5bbb953ffbd2a20a333704217c7077a78f96dfe5/console/src/app/pages/users/user-detail/auth-user-detail/edit-dialog/edit-dialog.component.ts#L14. You can then reference the type in the constructor of the dialog: https://github.com/zitadel/zitadel/blob/5bbb953ffbd2a20a333704217c7077a78f96dfe5/console/src/app/pages/users/user-detail/auth-user-detail/edit-dialog/edit-dialog.component.ts#L52. Opening the dialog is done like this: https://github.com/zitadel/zitadel/blob/5bbb953ffbd2a20a333704217c7077a78f96dfe5/console/src/app/pages/users/user-detail/auth-user-detail/auth-user-detail.component.ts#L399. This would be a great improvement

conblem commented on console/src/app/pages/users/user-detail/auth-user-detail/auth-user-mfa/auth-user-mfa.component.ts:
```diff
@@ -75,45 +85,33 @@ export class AuthUserMfaComponent implements OnInit, OnDestroy {
     this.service
       .listMyMultiFactors()
```
We are trying to move our api calls over to our @zitadel/client package. Could you add the listMyMultiFactors to the NewAuthService and use that call instead here: https://github.com/zitadel/zitadel/blob/5bbb953ffbd2a20a333704217c7077a78f96dfe5/console/src/app/services/new-auth.service.ts#L16

conblem commented on console/src/app/pages/users/user-detail/auth-user-detail/auth-user-mfa/auth-user-mfa.component.ts:
```diff
@@ -75,45 +85,33 @@ export class AuthUserMfaComponent implements OnInit, OnDestroy {
-      this.dataSource.data.splice(email, 1);
-    }
+  public applyOrgPolicy(): void {
+    this.service.getMyLoginPolicy().then((resp) => {
```
It would also be nice to use the NewAuthService here

conblem reviewed (commented):
I took a first look at your PR i'll go more in depth as soon as the requested changes are done as this will probably change the code quite a bit

AnthonyKot commented:
Added requested changes and also moved a couple of methods to new-service. Proof of work:
<img width="1505" alt="Screenshot 2025-02-22 at 14 53 38" src="https://github.com/user-attachments/assets/e8a7fa52-4da5-4997-97b3-b15cb65da1db" />

conblem commented:
Great I'll take a look at it maybe today latest tommorow.

AnthonyKot commented:
@conblem I am not in hurry :) sorry for huge change. it already contains 3 things:
1. fixing original bug (FE part)
2. make unit tests "compilable" + specific test
3. refactoring for v2 API compitability

conblem commented on console/src/app/pages/users/user-detail/auth-user-detail/auth-user-mfa/auth-user-mfa.component.spec.ts:
```diff
@@ -1,24 +1,115 @@
+import { MatDialog } from '@angular/material/dialog';
+import { NewAuthService } from 'src/app/services/new-auth.service';
+import { SecondFactorType } from 'src/app/proto/generated/zitadel/policy_pb';
+import { CardComponent } from '../../../../../modules/card/card.component';
```
pls don't use relative paths instead use src/app

conblem commented on console/src/app/pages/users/user-detail/auth-user-detail/auth-user-mfa/auth-user-mfa.component.ts:
```diff
@@ -30,40 +30,51 @@ export class AuthUserMfaComponent implements OnInit, OnDestroy {
   public AuthFactorState: any = AuthFactorState;
 
   public error: string = '';
+  public otpAvailable$ = new BehaviorSubject<boolean>(false);
```
can you change the public properties which are only used in the html to protected and also add readonly where possible

conblem commented on console/src/app/pages/users/user-detail/auth-user-detail/auth-user-mfa/auth-user-mfa.component.ts:
```diff
@@ -30,40 +30,51 @@ export class AuthUserMfaComponent implements OnInit, OnDestroy {
 
   constructor(
-    private service: GrpcAuthService,
+    private service: NewAuthService,
```
Readonly here as well

conblem reviewed (commented):
The changes are looking very good thanks for implementing the first changes I added some additional comments but I think this will soon be ready to merge. Maybe remerge the master back into your branch aswell just to make sure this won't be a problem in the future.

AnthonyKot commented:
All done. + adusted the test for the last changes:
<img width="621" alt="Screenshot 2025-02-26 at 21 18 21" src="https://github.com/user-attachments/assets/74d7e08a-2ede-4cae-9fb7-847025f3b309" />

AnthonyKot commented:
Hey, @conblem, needs to be re-approved once been merged on conflict in  'new-auth.service.ts' been resolved

conblem commented:
Sorry for the late reply I think tommorow I should have time to get this issue merged

muhlemmer commented:
Ping @conblem
